### PR TITLE
Allow creating nested folders in the file manager

### PR DESCRIPTION
**Pull request by leofeyer**
Implements contao/core#2907

@ausi Any security concerns?

### CODE REVIEW

**Comment by aschempp** [comment]
what if you enter `foo/../bar` in your demo setup?

**Comment by leofeyer** [comment]
It will not get past `Validator::isInsecurePath()`.

File: core-bundle/src/Resources/contao/dca/tl_files.php [hunk 1]
```diff
@@ -607,8 +607,9 @@ public function addFileLocation($value, DataContainer $dc)
 	public function checkFilename($varValue, DataContainer $dc)
 	{
 		$varValue = str_replace('"', '', $varValue);
+		$varValue = trim($varValue, '/');
```
Comment (ausi) [inline core-bundle/src/Resources/contao/dca/tl_files.php]: I think we should probably split the value here on `/` and validate every single part with the code below (not allowing `.` at the and and checking maxlength)

Comment (leofeyer) [reply core-bundle/src/Resources/contao/dca/tl_files.php]: Changed in 2c037b1e716c692bc469ac15b48eaac617b766c9.

Comment (ausi) [reply core-bundle/src/Resources/contao/dca/tl_files.php]: The maxlength check is still on the whole path, but it should be checked on every single chunk I think?

Comment (leofeyer) [reply core-bundle/src/Resources/contao/dca/tl_files.php]: Changed in 37e60da26f54a1f9c88dc333005e3eef0041be95.

**Comment by leofeyer** [comment]
I have just noticed another problem: The `checkFilename()` changes now also allow slashes when **renaming** files or folders. 🤔

**Comment by ausi** [comment]
> I have just noticed another problem: The `checkFilename()` changes now also allow slashes when **renaming** files or folders. 🤔

This can be seen as a feature I think. I often use this in IDEs to move files into a new subfolder.

**Review by ausi — DISMISSED** [review]

**Comment by aschempp** [comment]
> I have just noticed another problem: The `checkFilename()` changes now also allow slashes when **renaming** files or folders. 🤔

what happens if you rename a file and include slashes? creating subdirectories seems unintiutive to me?

**Comment by ausi** [comment]
> what happens if you rename a file and include slashes? creating subdirectories seems unintiutive to me?

Does it? I always use this in code editors to move files to non-yet-existing subdirectories...

**Comment by leofeyer** [comment]
> This can be seen as a feature I think. I often use this in IDEs to move files into a new subfolder.

I agree, it just does not work yet. 😄

**Comment by Aybee** [comment]
> what happens if you rename a file and include slashes? creating subdirectories seems unintiutive to me?

I also use this technique in filezilla.

**Comment by leofeyer** [comment]
Turns out we cannot modify the folder structure when renaming existing resources.

```
files/
    foo/
        bar/
            baz/
```

When trying to rename the `files/foo/bar` folder to `bar/bar`, the following exception is thrown:



Therefore I have adjusted the code to only allow using nested folders when creating new folders (as initially requested).

**Review by Toflar — APPROVED** [review]

**Pull request merged**